### ISSUE TITLE
Feature Request: Ability to maintain stable style prop references in React Native

### ISSUE THREAD

**Issue by philipbulley** [styled-components/styled-components#3149]
## Problem
Currently, in `styled-components/native` the `style` prop passed into a styled custom component is not stable — a new object reference is passed in on every render, regardless of whether any of the styling has changed.

#### Example
https://snack.expo.io/@pbcodes/7e9d69
Please run this snack using the Android or iOS devices. The issue doesn't apply to React Native Web.

## Motivation
Stable object references are a fundamental part of React and its performant application. We should make provisions for this wherever possible.

Without this ability, we cannot make optimisations using `React.memo` in an idiomatic way.

## Prior Art
This problem was raised in #3020 by @Federkun but was ultimately closed with only a workaround suggested.

@kitten correctly [pointed out](https://github.com/styled-components/styled-components/pull/3020#issuecomment-585487853) that `styled(CustomComponent)` should be used sparingly, however, this doesn't mean that this usage is wrong. The forwarding of the `style` prop can often be a useful pattern in adding layout properties such as `margin` from the outside of a custom component without having to create wrapper containers.

In lieu of stable object references, it was also suggested that `React.memo` could be used in conjunction with a comparator function, such as `isEqual` from `react-fast-compare`. However, there is a separation of concerns downside to this as any custom component forwarding the `style` prop and also exposed via `React.memo` must assume that there is a possibility that `styled-components` may be used and thus preemptively employ a deep equality check. If the custom component is part of a shared component library, it is unreasonable for the component to expect to be styled via `styled-components`. In cases where `StyleSheet` is used, there will be an unnecessary baked-in performance hit to the custom component via its usage of `isEqual`.

## Conslusion
I think it's reasonable to ask that `styled-components` itself should provide support for the fundamental ability to work with stable object references on an opt-in basis.

**Comment by philipbulley** [comment]
## Proposal
Instead of forcing all components to take on the overhead of memoization, can we export an additional form of `styled` as:

```js
import { styledMemo } from 'styled-components/native';
import CustomComponent from './CustomComponent';

const CustomComponentStyled = styled(CustomComponent)`
...
`;
```

I've created a WIP PR #3150 exploring a way that this might be achieved.

**Comment by probablyup** [comment]
I think we'll look into this after https://github.com/styled-components/styled-components/pull/3152 is merged

**Comment by philipbulley** [comment]
@probablyup Related PR https://github.com/styled-components/styled-components/pull/3150 — would be great to get your review (and possibly advise on the flowtype def issue, should be an easy one). Thanks.

**Comment by kitten** [comment]
Closing due to inactivity.

This isn't really an issue I think we should resolve, unless we make the React Native version a separate package of the mainline one. See: #1617
